services/bundle_analysis/new_notify/contexts/__init__.py

```diff
@@ -0,0 +1,152 @@
+from functools import cached_property
+
+from shared.bundle_analysis import (
+    BundleAnalysisReport,
+    BundleAnalysisReportLoader,
+)
+from shared.torngit.base import TorngitBaseAdapter
+from shared.yaml import UserYaml
+
+from database.enums import ReportType
+from database.models.core import Commit, Repository
+from database.models.reports import CommitReport
+from services.archive import ArchiveService
+from services.bundle_analysis.new_notify.types import NotificationType
+from services.repository import (
+    get_repo_provider_service,
+)
+from services.storage import get_storage_client
+
+
+class BaseBundleAnalysisNotificationContext:
+    """Base NotificationContext for bundle analysis notifications.
+    It includes basic information that all bundle analysis notifications need.
+    Use NotificationContextBuilder to populate the context.
+
+    Example:
+      builder = NotificationContextBuilder(commit, current_yaml, GITHUB_APP_INSTALLATION_DEFAULT_NAME)
+      notification_context = builder.build_context().get_result()
+    """
+
+    notification_type: NotificationType
+
+    def __init__(
+        self, commit: Commit, current_yaml: UserYaml, gh_app_installation_name: str
+    ) -> None:
+        self.commit = commit
+        self.current_yaml = current_yaml
+        self.gh_app_installation_name = gh_app_installation_name
+
+    @cached_property
+    def repository(self) -> Repository:
+        return self.commit.repository
+
+    @cached_property
+    def repository_service(self) -> TorngitBaseAdapter:
+        return get_repo_provider_service(
+            self.repository,
+            installation_name_to_use=self.gh_app_installation_name,
+        )
+
+    @property
+    def commit_report(self) -> CommitReport:
+        return self._commit_report
+
+    @commit_report.setter
+    def commit_report(self, report: CommitReport) -> None:
+        self._commit_report = report
+
+    @property
+    def bundle_analysis_report(self) -> BundleAnalysisReport:
+        return self._bundle_analysis_report
+
+    @bundle_analysis_report.setter
+    def bundle_analysis_report(self, report: BundleAnalysisReport) -> None:
+        self._bundle_analysis_report = report
+
+
+class NotificationContextBuildError(Exception):
+    def __init__(self, failed_step: str, detail: str | None = None) -> None:
+        super()
+        self.failed_step = failed_step
+        self.detail = detail
+
+
+class WrongContextBuilderError(Exception):
+    pass
+
+
+class NotificationContextBuilder:
+    """Creates the BaseBundleAnalysisNotificationContext one step at a time, in the correct order."""
+
+    # TODO: Find a way to re-use the base-context (by cloning it instead of recreating it for every notification)
```

Comment:
Could we try a singleton pattern approach perhaps? Make it once, save it, and if it exists already, use it?

Reply:
Singleton I don't think was the answer.
I did something tho, in which you can initialize the builder from an existing context by looking at what fields the existing context has and porting them over to the new one.

Because notification is a read-only process it's OK to have the same reference in multiple places.

Reply:
Huh, does this have a pattern?? Not really familiar w/ this